rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,501 @@
+  - There are more complicated things.
+
+-> Answer: A Tool might be able to keep track of a percentage of expressions, and sometimes it may be very hard or impossible. For that case, the doc-comment can offer a dedicated Keyword to override the scope.
+
+e.g.
+
+The following is an idea for a problem that will arise if tools try to track doc-comments' positions and the location in the nixpkgs tree. (Although this problem is not nixpkgs specific)
+
+```nix
+## This file is called somewhere that cannot be automatically tracked/is impossible to analyze statically.
+## The 'TreePath' override can be used by the docstring author to set fixed paths in the nixpkgs expression.
+## This tells the doc-tool that the expression is available in two aliases: `pkgs.stdenv` and `lib.sth` 
+## (This behavior will not be specified in this RFC)
+## # Meta
+## ## TreePath
+## - pkgs.stdenv
+## - lib.sth
+
+{...}:
+{
+    # returns something
+}
+```
+
+# Future work
+[future]: #future-work
+
+## Editor support
+
+When starting hitting {enter} inside a doc-block the new line, should be automatically prefixed with `##` or `#|` accordingly.
+This is done in rust similarly. Nix already offers a bunch of LSP's e.g. [nil](https://github.com/oxalica/nil), [rnix-lsp](https://github.com/nix-community/rnix-lsp) are the most common ones.
+Those LSP's should implement the simple "line continuation" feature. (I don't know the exact name here)
+
+## Nixodc
+
+Nixdoc needs to be changed in order to differentiate between regular comment and doc-comments.
+There might be an intermediate phase of transition, where the old syntax and features is supported for a while.
+
+- When extending nixdoc or writing dedicated parsers, the following persons can assist: [@hsjobeki]
+
+## Documentation generators
+
+Generating documentation from doc-comments is still a challenge.
+If we'd like the fully automated approach, we definitely need something that can also evaluate nix expressions. 
+(We have such a module in `tvix` which needs to be investigated more here)
+
+Alternatively we can use the future nixdoc together with a static `map.json` that contains the scope for every discovery file/path in nixpkgs.
+
+As this second approach is much easier I propose this is how we should initially start to extend the scope.
+
+## More specialized section headings
+
+### Type
+
+- An RFC under construction specifies the used syntax within the `Type`-Block. It depends on this RFC, as it is the groundwork to provide a standardized field where additional rules can apply.
+
+## Native support in nix
+
+- `NixOS/nix` should implement native support for doc-comments so that our users don't have to rely on nixpkgs or external tools. Those tools can still exist and provide more custom functionality, but documenting nix expressions should be natively possible.
```

Comment:
I am more minimalistic in this specific point.

First, because Nix is now a "bundled multitool", similar to busybox (where `cp == busybox cp`, `mount == busybox mount` etc.). 

What "implement native support" means here? There should be a `nix doc export fancyFile.pdf`? Or `nix repl` should have a special `showDocs` function to read docblocks?

Second, what if I want to use it in a constrained cloud à la GitHub Actions? Can I compile nix without doc support?

Reply:
Should I delete this point? I don't want to specify what "native support" means. I thought of something like the "nix repl doc" feature from the PR I mentioned earlier. Or, more generally, doc support in the native Nix tooling you get if you installed just nix: that could mean either previewing rendered documentation in some way or retrieving doc-comments from specific items. (Without the need to install any extra tools)

Reply:
I believe RFCs should be focused on only one thing. It helps focusing on the problem to be solved.

Defining a format (grammar, semantics etc.) of in-code documentation is one thing.

Defining if we should modify the nix evaluator to parse this in-code documentation is another thing.